File: maths/L04Induction.java
public class L04Induction {
  public static void main(String[] args) {
    int grid = 63;
    long start, end = 0;
    long st, en = 0;

    start = System.currentTimeMillis();
    System.out.println(String.format("%d", (long)(Math.pow(2, grid)) - 1));
    end = System.currentTimeMillis();
    System.out.println(String.format("%d", (end - start)));

    st = System.currentTimeMillis();
    System.out.println(String.format("%d", getNumberOfWheat(grid)));
    en  = System.currentTimeMillis();
    System.out.println(String.format("%d", (end - start)));

    grid = 64;
    Result result = new Result();
    System.out.println(prove(grid, result));

    st = System.currentTimeMillis();
    System.out.println(String.format("%d", L03Iterative.getNumberOfWheat(grid)));
    en  = System.currentTimeMillis();
    System.out.println(String.format("%d", (end - start)));
  }

  public static long getNumberOfWheat(int grid) {
    long sum = 0;
    long numberOfWheatInGrid = 0;
    numberOfWheatInGrid = 1;
    sum += numberOfWheatInGrid;
    for (int i = 2; i <= grid; i++) {
      numberOfWheatInGrid *= 2;
      sum += numberOfWheatInGrid;
    }
    return sum;
  }

  public static boolean prove(int k, Result result) {
    // 证明 n = 1 时，命题是否成立
    if (k == 1) {
      if((Math.pow(2, 1) - 1) == 1) {
        result.wheatNum = 1;
        result.wheatTotalNum = 1;
        return true;
      } else {
        return false;
      }
    }
    // 如果 n = (k-1) 时命题成立，证明 n = k 时命题是否成立
    else {
      boolean proveOfPreviousOne = prove(k-1, result);
      result.wheatNum *= 2;
      result.wheatTotalNum += result.wheatNum;
      boolean proveOfCurrentOne = false;
      if (result.wheatTotalNum == (Math.pow(2, k) - 1)) {
        proveOfCurrentOne = true;
      }
      if (proveOfPreviousOne && proveOfCurrentOne) {
        return true;
      } else {
        return false;
      }
    }
  }
}

class Result {
    // 当前格的麦粒数
    public long wheatNum = 0;
    // 目前为止麦粒的总数
    public long wheatTotalNum = 0;
}

File: maths/L04Induction.py
class Result:
    current_wheat = 1
    total_wheat = 1

def proveWithPow(k, result):
    if k == 1:
        if (pow(2, 1) - 1) == 1:
            return True
        else:
            return False
    else:
        provePrevious = proveWithPow(k-1, result)
        result.current_wheat *= 2
        result.total_wheat += result.current_wheat
        proveCurrent = False
        if (pow(2,k)-1) == result.total_wheat:
            proveCurrent = True
        if provePrevious & proveCurrent:
            return True
        else:
            return False

def get_result(n):
    return 1 if n==1 else get_result(n-1)*2 + 1

def prove(k, result):
    if k == 1:
        if  (2 ** 1 - 1) == 1:
            result.current_wheat = 1
            result.total_wheat = 1
            return True
        else:
            return False
    else:
        provePrevious = prove(k-1, result)
        result.current_wheat *= 2
        result.total_wheat += result.current_wheat
        proveCurrent = False
        if result.total_wheat == (2**k - 1):
            proveCurrent = True
        if provePrevious & proveCurrent:
            return True
        else:
            return False
    return

if __name__ == '__main__':
    result = Result()
    print(proveWithPow(53, result))
    print(proveWithPow(54, result))

    print(get_result(63))

    grid = 63
    result = Result()
    print(prove(grid, result))
    print(prove(64, result))
    print(prove(127, result))
    print(prove(999, result))
